frontend/app/controllers/spree/products_controller.rb

```diff
@@ -3,13 +3,16 @@ class ProductsController < Spree::StoreController
     include Spree::ProductsHelper
 
     before_action :load_product, :load_variants, only: :show
-    before_action :load_taxon, :load_products, only: :index
+    before_action :load_taxon, only: :index
 
     helper 'spree/taxons'
 
     respond_to :html
 
     def index
+      @searcher = build_searcher(params.merge(include_images: true))
+      @products = @searcher.retrieve_products
+      @products = @products.includes(:possible_promotions) if @products.respond_to?(:includes)
```

Comment:
Isn't `@products` always an ActiveRecord relation that always responds to `#includes`?

Reply:
@KirillKayumov not necessary as you can swap the searcher class for something different. This is the way it was in previous spree versions and we need to keep it so all extensions/customizations work properly